src/client/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs

```diff
@@ -1240,14 +1329,31 @@ bool ITokenCacheInternal.HasTokensNoLocks()
                 return;
             }
 
+            //Filter and remove tokens based on account identifier as 
             string partitionKey = account.HomeAccountId.Identifier;
 
+            ILoggerAdapter logger = requestContext.Logger;
+
             var refreshTokens = Accessor.GetAllRefreshTokens(partitionKey);
-            refreshTokens.RemoveAll(item => !item.HomeAccountId.Equals(account.HomeAccountId.Identifier, StringComparison.OrdinalIgnoreCase));
+            refreshTokens.RemoveAll(item => !item.HomeAccountId.Equals(partitionKey, StringComparison.OrdinalIgnoreCase));
+
+            RemoveRefreshTokens(refreshTokens, logger, out bool filterByClientId);
+
+            var accessTokens = Accessor.GetAllAccessTokens(partitionKey);
+            accessTokens.RemoveAll(item => !item.HomeAccountId.Equals(partitionKey, StringComparison.OrdinalIgnoreCase));
```

Comment:
The three should be like a transaction. What if one of the methods fails? That can lead to inconsistency.

Reply:
This is the original logic for the remove account method and I am assuming that if one the removal methods fail, then we assume the cache is not formatted properly. Otherwise, we should try to remove everything we can from the cache and if there comes a point where something is missing, the developer can always reacquire a new set of tokens.